docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+not be caught by any per-timeline scrubbing (see [scrubbing](#cleaning-up-orphan-objects-scrubbing)), since scrubbing happens on the
+attached pageserver, and once the timeline is deleted it isn't attached anywhere.
+This scenario should be pretty rare, and the control plane can make it even
+rarer by ensuring that if a tenant is in a multi-attached state (e.g. during
+migration), we wait for that to complete before processing the deletion. Beyond
+that, we may implement some other top-level scrub of timelines in
+an external tool, to identify any tenant/timeline paths that are not found
+in the control plane database.
+
+Examples:
+
+- Deletion, node restarts partway through:
+  - By the time we returned 204, we have written a remote delete marker
+  - Any subsequent incarnation of the same node_id will see the remote
+    delete marker and continue to process the deletion
+  - We only require that the control plane does not change the tenant's
+    attachment while waiting for the deletion.
+  - If the original pageserver is lost permanently and no replacement
+    with the same node_id is available, then the control plane must recover
+    by re-attaching the tenant to a different node.
+- Creation, node becomes unresponsive partway through.
+  - Control plane will see HTTP request timeout, keep re-issuing
+    request to whoever is the latest attachment point for the tenant
+    until it succeeds.
+  - Stale nodes may be trying to execute timeline creation: they will
+    write out index_part.json files with stale node generation or
+    stale attachment generation: these will be eventually cleaned up
+    by the same mechanism as other old indices.
+
+## Implementation Part 2: Optimizations
+
+### Persistent deletion queue
+
+Between writing our a new index_part.json that doesn't reference an object,
+and executing the deletion, an object passes through a window where it is
+only referenced in memory, and could be leaked if the pageserver is stopped
+uncleanly. That introduces conflicting incentives: we would like to delay
+deletions to minimize control plane load and allow aggregation of full-sized
+DeleteObjects requests, but we would also like to minimize leakage by executing
+deletions promptly.
+
+To resolve this, we may make the deletion queue persistent, writing out
+deletion lists as S3 objects for later execution. This shrinks the window
+of possible object loss to the gap between writing index_part.json, and
+writing the next deletion list ot S3. The actual deletion of objects
+(with its requirement to sync with the control plane) may happen much
+later.
+
+The flow of deletion becomes:
+
+1. Enqueue in memory
+2. Enqueue persistently by writing out a deletion list, storing the
+   attachment generations and node generation.
+3. Validate the deletion list by calling to the control plane
+4. Execute the valid parts of the deletion list (i.e. call DeleteObjects)
```

Comment:
It's not clear whether the deletion queue is shared between node generations or not.

And I think you need to make it clearer how exactly the bulk validation works, and what "valid parts" means exactly.

My understanding so far is that we have a deletion queue per node generation.
Within a deletion queue for node generation X are entries of the shape `{ KeyToDelete, AttachmentGenerationThatTheKeyIsPartOf }`.
The processing of the delete queue is like:

1. Pop entry off queue
2. Validate with control plane: is suffix `X-AttachementGenerationThatTheKeyIsPartOf` still the most recent for the given tenant.
3. If yes: carry out the deletion.
4. If no: skip the entry and move on to the next one.

Correct?

Reply:
Further down, you lay out the proposal for C. I think for the sake of clarity, it makes sense to commit to Option C, so that you can remove all the stuff between here and the `Deletion queue persistence format` section.